### PR TITLE
xFormers attention op arg

### PR DESCRIPTION
**Pull request by takuma104**
# What does this PR do?
To add the `attention_op` argument to `enable_xformers_memory_efficient_attention()`. This argument can override the `op` argument of `memory_efficient_attention()` in xFormers. It was [originally written](https://github.com/huggingface/diffusers/commit/ae0cc0b71f28c0f2c5c27026b18f1bea98b505f1) by @patil-suraj on the `xformers-attention-op-arg` branch, and I have added some tweaks so that it can be merged into the current main branch. A short documentation has also been added.

# Usage Example:
```python
import xformers
import xformers.ops

op = xformers.ops.MemoryEfficientAttentionFlashAttentionOp
pipe.enable_xformers_memory_efficient_attention(attention_op=op)
```

As an example of the application of this PR, The use of Flash Attention improves the reproducibility of Stable Diffusion image generation due to its deterministic behavior. Discussed in #1997.

I am confirming this PR code by using the following code.
https://gist.github.com/takuma104/acc9ff3809e4b259bf24b8130c021823

### CODE REVIEW

**Review by patil-suraj — APPROVED** [review]
Very cool, and looks good to me. Thanks a lot for working on this!

File: src/diffusers/pipelines/pipeline_utils.py [hunk 1]
```diff
@@ -847,22 +847,30 @@ def enable_xformers_memory_efficient_attention(self):
 
         Warning: When Memory Efficient Attention and Sliced attention are both enabled, the Memory Efficient Attention
         is used.
+
+        Parameters:
+            attention_op (`Callable`, *optional*):
+                Override the default `None` operator for use as `op` argument to the
+                [`memory_efficient_attention()`](https://facebookresearch.github.io/xformers/components/ops.html#xformers.ops.memory_efficient_attention)
+                function of xFormers.
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/pipeline_utils.py:855]: Could we maybe add an example here of how to pass a specific `attention_op`, e.g.:

```suggestion
                function of xFormers.
                
        Examples:
        ```py
        from diffusers import DiffusionPipeline
        from xformers import ...  # some attention op
        
        pipe = DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
        pipe.enable_xformers_memory_efficient_attention(attention_op=...)
        ```
```

File: src/diffusers/models/modeling_utils.py [hunk 2]
```diff
@@ -214,8 +216,14 @@ def enable_xformers_memory_efficient_attention(self):
 
         Warning: When Memory Efficient Attention and Sliced attention are both enabled, the Memory Efficient Attention
         is used.
+
+        Parameters:
+            attention_op (`Callable`, *optional*):
+                Override the default `None` operator for use as `op` argument to the
+                [`memory_efficient_attention()`](https://facebookresearch.github.io/xformers/components/ops.html#xformers.ops.memory_efficient_attention)
+                function of xFormers.
```
Comment (patrickvonplaten) [inline src/diffusers/models/modeling_utils.py:224]:
```suggestion
                function of xFormers.
                
        ```suggestion
                function of xFormers.
                
        Examples:
        ```py
        from diffusers import UNet2DConditionModel
        from xformers import ...  # some attention op
        
        model = UNet2DConditionModel.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="unet")
        model.enable_xformers_memory_efficient_attention(attention_op=...)
        ```
```

Comment (patrickvonplaten) [reply src/diffusers/models/modeling_utils.py:224]: One example here would be nice

**Review by pcuenca — COMMENTED** [review]
I agree with Patrick, a couple of examples would be nice.

**Comment by takuma104** [comment]
thanks @patrickvonplaten! I've merged examples.

**Comment by takuma104** [comment]
hmm... I've got the following docstring error. The `...` is wrong? or something?
`Error message:
Cannot parse: 2:21: from xformers import ...  # some attention op`

File: src/diffusers/models/modeling_utils.py [hunk 3]
```diff
@@ -214,8 +216,23 @@ def enable_xformers_memory_efficient_attention(self):
...
+
+        Examples:
+        ```py
+        from diffusers import UNet2DConditionModel
+        from xformers import ...  # some attention op
+
+        model = UNet2DConditionModel.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="unet")
+        model.enable_xformers_memory_efficient_attention(attention_op=...)
+        ```
```
Comment (patil-suraj) [inline src/diffusers/models/modeling_utils.py:233]:
```suggestion
        Examples:

        ```py
        >>> from diffusers import UNet2DConditionModel
        >>> from xformers.ops import MemoryEfficientAttentionFlashAttentionOp

        >>> model = UNet2DConditionModel.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="unet").to("cuda")
        >>>  model.enable_xformers_memory_efficient_attention(attention_op=MemoryEfficientAttentionFlashAttentionOp)
        ```
```

File: src/diffusers/pipelines/pipeline_utils.py [hunk 4]
```diff
@@ -847,22 +847,39 @@ def enable_xformers_memory_efficient_attention(self):
...
+
+        Examples:
+        ```py
+        from diffusers import DiffusionPipeline
+        from xformers import ...  # some attention op
+
+        pipe = DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
+        pipe.enable_xformers_memory_efficient_attention(attention_op=...)
+        ```
```
Comment (patil-suraj) [inline src/diffusers/pipelines/pipeline_utils.py:864]:
```suggestion
        Examples:

        ```py
        >>> from diffusers import DiffusionPipeline
        >>> from xformers.ops import MemoryEfficientAttentionFlashAttentionOp

        >>> pipe = DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4").to("cuda")
        >>> pipe.enable_xformers_memory_efficient_attention(attention_op=MemoryEfficientAttentionFlashAttentionOp)
        ```
```

**Comment by patil-suraj** [comment]
Hey @takuma104 , thanks for adding the examples. The docstring error was because of the wrong syntax (`...`), the docstrings expects a valid python example. I  left suggestions for how to fix this in the comments.

**Comment by takuma104** [comment]
Thanks for your help, @patil-suraj ! Wow, that's a very strict code-style checker. Nice.

**Comment by takuma104** [comment]
I ran the example code to make sure, Flash Attention does not work well with SD1.4, but with SD2.1. Can I change "CompVis/stable-diffusion-v1-4" to "stabilityai/stable-diffusion-2-1"?

**Comment by takuma104** [comment]
Which one would be more appropriate as a code example?
- It does not work as code, but as an example it is simple and understandable. The only change is "CompVis/stable-diffusion-v1-4” to `model_id`. 
```py
from diffusers import DiffusionPipeline
from xformers.ops import MemoryEfficientAttentionFlashAttentionOp
pipe = DiffusionPipeline.from_pretrained(model_id).to("cuda")
pipe.enable_xformers_memory_efficient_attention(attention_op=MemoryEfficientAttentionFlashAttentionOp)
```
- Fully functional code:
```py
import torch
from diffusers import DiffusionPipeline
from xformers.ops import MemoryEfficientAttentionFlashAttentionOp
pipe = DiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-2-1", torch_dtype=torch.float16).to("cuda")
pipe.enable_xformers_memory_efficient_attention(attention_op=MemoryEfficientAttentionFlashAttentionOp)
# Workaround for not accepting attention shape using VAE for Flash Attention
pipe.vae.enable_xformers_memory_efficient_attention(attention_op=None)
```

**Comment by patil-suraj** [comment]
Hey @takuma104 the second example looks great, it's always better to have fully functional examples in docs so readers can just copy-paste it to try. 

Also out of curiosity, what do you mean by Flash Attention does not work well with SD1.4, think it works no ?

**Comment by takuma104** [comment]
Hi @patil-suraj 

Ok, thanks. I got it. I'll update it very soon.

>Also out of curiosity, what do you mean by Flash Attention does not work well with SD1.4, think it works no ?

The latest xFormers show the error in detail, and when I do U-Net inference with SD1.x, I get the following.

```
ValueError: Operator `memory_efficient_attention` does not support inputs:
     query       : shape=(16, 256, 1, 160) (torch.float16)
     key         : shape=(16, 256, 1, 160) (torch.float16)
     value       : shape=(16, 256, 1, 160) (torch.float16)
     attn_bias   : <class 'NoneType'>
     p           : 0.0
`flshattF` is not supported because:
    max(query.shape[-1] != value.shape[-1]) > 128
```   
In SD1.x, it seems that some of the U-Net attentions' size of K in [B, M, K] doesn't fit the required. [It must be less than 128. ](https://github.com/facebookresearch/xformers/blob/main/xformers/ops/fmha/flash.py#L198)

In SD2.x, I don't know if this is intentional or not, but it seems that the size of K is 128 or less in all cases.

**Comment by patil-suraj** [comment]
Thanks for updating the doc!

**Comment by patil-suraj** [comment]
@adhikjoshi  Feel free to open an issue with a reproducible code snippet for this :)

**Comment by tianleiwu** [comment]
I also encountered same error (max(query.shape[-1] != value.shape[-1]) > 128') in T4 GPU even with xformers 0.0.16 and 0.0.17.dev444 on stable diffusion 1.5 model.

Is it supported only in Amper or later GPU (with cuda compute capactity >= 80)?

**Comment by takuma104** [comment]
Hi @tianleiwu,

Want to use Flash Attention? [As mentioned above](https://github.com/huggingface/diffusers/pull/2049#issuecomment-1402178059), SD1.x models are not supported with Flash Attention because they don't meet the dim<=128 requirement. SD 2.x models seem to be supported with Flash Attention.

[According to the xFormers code](https://github.com/facebookresearch/xformers/blob/main/xformers/ops/fmha/flash.py#L191), Tesla T4=sm_75 is in the minimum requirement, so Tesla T4 is supported with Flash Attention.

**Comment by adhikjoshi** [comment]
> Hi @tianleiwu,
> 
> Want to use Flash Attention? [As mentioned above](https://github.com/huggingface/diffusers/pull/2049#issuecomment-1402178059), SD1.x models are not supported with Flash Attention because they don't meet the dim<=128 requirement. SD 2.x models seem to be supported with Flash Attention.
> 
> [According to the xFormers code](https://github.com/facebookresearch/xformers/blob/main/xformers/ops/fmha/flash.py#L191), Tesla T4=sm_75 is in the minimum requirement, so Tesla T4 is supported with Flash Attention.

Actually, it works with flash attention. I tried with all models. Works great on 3090.

Only img2img and inpainting doesn't work. Hence had to drop it.

**Pull request merged**